### PR TITLE
SubAdmin: Null handling

### LINKED ISSUE #20070: PHP fatal in subadmin
@rullzer 

````
Mon Oct 26 22:06:54 2015] PHP Fatal error:  Call to a member function getUID() on null in /home/deepdiver/Development/ownCloud/master/lib/private/subadmin.php on line 190
[Mon Oct 26 22:06:54 2015] PHP Stack trace:
[Mon Oct 26 22:06:54 2015] PHP   1. {main}() /home/deepdiver/Development/ownCloud/master/index.php:0
[Mon Oct 26 22:06:54 2015] PHP   2. OC::handleRequest() /home/deepdiver/Development/ownCloud/master/index.php:39
[Mon Oct 26 22:06:54 2015] PHP   3. OC\Route\Router->match() /home/deepdiver/Development/ownCloud/master/lib/base.php:852
[Mon Oct 26 22:06:54 2015] PHP   4. call_user_func:{/home/deepdiver/Development/ownCloud/master/lib/private/route/router.php:282}() /home/deepdiver/Development/ownCloud/master/lib/private/route/router.php:282
[Mon Oct 26 22:06:54 2015] PHP   5. OC\AppFramework\routing\RouteActionHandler->__invoke() /home/deepdiver/Development/ownCloud/master/lib/private/route/router.php:282
[Mon Oct 26 22:06:54 2015] PHP   6. OC\AppFramework\App::main() /home/deepdiver/Development/ownCloud/master/lib/private/appframework/routing/routeactionhandler.php:45
[Mon Oct 26 22:06:54 2015] PHP   7. Pimple\Container->offsetGet() /home/deepdiver/Development/ownCloud/master/lib/private/appframework/routing/routeactionhandler.php:100
[Mon Oct 26 22:06:54 2015] PHP   8. OC\AppFramework\DependencyInjection\DIContainer->OC\AppFramework\DependencyInjection\{closure}() /home/deepdiver/Development/ownCloud/master/3rdparty/pimple/pimple/src/Pimple/Container.php:113
[Mon Oct 26 22:06:54 2015] PHP   9. Pimple\Container->offsetGet() /home/deepdiver/Development/ownCloud/master/3rdparty/pimple/pimple/src/Pimple/Container.php:265
[Mon Oct 26 22:06:54 2015] PHP  10. OC\AppFramework\DependencyInjection\DIContainer->OC\AppFramework\DependencyInjection\{closure}() /home/deepdiver/Development/ownCloud/master/3rdparty/pimple/pimple/src/Pimple/Container.php:113
[Mon Oct 26 22:06:54 2015] PHP  11. Pimple\Container->offsetGet() /home/deepdiver/Development/ownCloud/master/3rdparty/pimple/pimple/src/Pimple/Container.php:312
[Mon Oct 26 22:06:54 2015] PHP  12. OC\Settings\Application->OC\Settings\{closure}() /home/deepdiver/Development/ownCloud/master/3rdparty/pimple/pimple/src/Pimple/Container.php:113
[Mon Oct 26 22:06:54 2015] PHP  13. OC\AppFramework\Utility\SimpleContainer->query() /home/deepdiver/Development/ownCloud/master/settings/application.php:169
[Mon Oct 26 22:06:54 2015] PHP  14. Pimple\Container->offsetGet() /home/deepdiver/Development/ownCloud/master/lib/private/appframework/utility/simplecontainer.php:104
[Mon Oct 26 22:06:54 2015] PHP  15. OC\Settings\Application->OC\Settings\{closure}() /home/deepdiver/Development/ownCloud/master/3rdparty/pimple/pimple/src/Pimple/Container.php:113
[Mon Oct 26 22:06:54 2015] PHP  16. OC_SubAdmin::isSubAdmin() /home/deepdiver/Development/ownCloud/master/settings/application.php:202
[Mon Oct 26 22:06:54 2015] PHP  17. OC\SubAdmin->isSubAdmin() /home/deepdiver/Development/ownCloud/master/lib/private/legacy/subadmin.php:143
[Mon Oct 26 22:06:54 2015] ::1:46908 [500]: /index.php/settings/apps - Call to a member function getUID() on null in /home/deepdiver/Development/ownCloud/master/lib/private/subadmin.php on line 190

### PR DESCRIPTION
**Pull request by DeepDiver1975**
fixes #20070

@rullzer @icewind1991 @PVince81 @MorrisJobke THX

### CODE REVIEW

File: lib/private/subadmin.php [hunk 1]
```diff
@@ -185,7 +185,10 @@ public function isSubAdminofGroup(IUser $user, IGroup $group) {
 	 * @param IUser $user 
 	 * @return bool
 	 */
-	public function isSubAdmin(IUser $user) {
+	public function isSubAdmin($user) {
```
Comment (MorrisJobke) [inline lib/private/subadmin.php]: I would love to know why it is null ;) Fixing the cause vs fixing the symptoms

Comment (DeepDiver1975) [inline lib/private/subadmin.php]: https://github.com/owncloud/core/blob/fix-subadmin-null-user/lib/private/legacy/subadmin.php#L143

Passing in null is always possible and has to be taken care of

Comment (LukasReschke) [inline lib/private/subadmin.php]: `IUser $user = null` might work then. Or we fix the code that calls this :wink:

Comment (DeepDiver1975) [inline lib/private/subadmin.php]: Fix many call locations or one implementation?

Seriously what are we dusgussing here?

**Comment by rullzer** [comment]
Ah yes of course the user and group manager do not throw exceptions but return null :(

Let me make fix the whole class then in this PR. Since else we are doomed to stumble upon this again in a slightly different context

**Comment by rullzer** [comment]
Now I'm sort of happy with it. We still enforce types but accept null.

Also now that code is tested by unit tests instead of intergration tests. YAY!

**Comment by DeepDiver1975** [comment]
hmmmm ... adding the default null is so wrong from my perspective.

Why setting a default value which is anyhow not valid/non-sense?

**Comment by MorrisJobke** [comment]
> Why setting a default value which is anyhow not valid/non-sense?

Then why do we pass in null if it is not of type `IUser`? That was also the question above: We should fix the cause (wrong usage of those methods) and not make the methods deal with this.

**Comment by rullzer** [comment]
Else we need a ton of extra checks... because what happens if a user passes an IGroup object instead of a IUser etc... I mean if we don't expect the caller to verify that the user is not null then we can't expect them to verify that it is an IUser either...

**Comment by DeepDiver1975** [comment]
- default null is bullshit because calling the method without any argument makes no sense
- because of what ever reason if we pass in null the instance totally falls apart
- fixing the caller sounds stupid because there will be many places where we have to duplicate the code

=> move the check into the method

**Comment by rullzer** [comment]
So like this then? I feel kind of sad that we throw the type safety away...

**Comment by DeepDiver1975** [comment]
> I feel kind of sad that we throw the type safety away...

I honestly don't get the intention of php here - null *to me* is a valid value for any object type ... anyway

**Comment by LukasReschke** [comment]
This kind of stuff type insecurity is what gave us security bugs such as the famous "User can access all files via share" as the return value was `null`. As this is fairly recent code in private API only I'm preparing another PR just adjusting the usage of code that calls this API. In my PoV it's the callers responsibility here to ensure type integrity.

**Comment by rullzer** [comment]
@LukasReschke yeah fixing OC_SubAdmin should be sufficient for this. And then we just make sure to fix all the calls properly when we confert to the non static stuff.

**Comment by DeepDiver1975** [comment]
> yeah fixing OC_SubAdmin should be sufficient for this.

please kill that class - nobody needs it

**Comment by LukasReschke** [comment]
> please kill that class - nobody needs it

That's another way to go. Ok. Let's do that then.

**Comment by blizzz** [comment]
In all the DB Queries in \OC\Subadmin, fetching the user and/or group follows. There null-return values should be sorted out just as well. Happens, for instance, if you (temporarily) disable an user/group backend.

**Comment by LukasReschke** [comment]
My proposal: https://github.com/owncloud/core/pull/20135

**Comment by MorrisJobke** [comment]
Replaced -> closing

**Pull request closed without merging**